src/workerd/api/sql.c++

```diff
@@ -115,24 +154,85 @@ double SqlStorage::Cursor::getRowsWritten() {
   }
 }
 
+SqlStorage::Cursor::RowIterator::Next SqlStorage::Cursor::next(jsg::Lock& js) {
+  KJ_IF_SOME(s, state) {
+    cachedColumnNames.ensureInitialized(js, s->query);
+  }
+
+  auto self = JSG_THIS;
+  auto maybeRow = rowIteratorNext(js, self);
+  bool done = maybeRow == kj::none;
+  return {
+    .done = done,
+    .value = kj::mv(maybeRow),
+  };
+}
+
+jsg::JsArray SqlStorage::Cursor::toArray(jsg::Lock& js) {
+  KJ_IF_SOME(s, state) {
+    cachedColumnNames.ensureInitialized(js, s->query);
+  }
+
+  auto self = JSG_THIS;
+  v8::LocalVector<v8::Value> results(js.v8Isolate);
+  for (;;) {
+    auto maybeRow = rowIteratorNext(js, self);
+    KJ_IF_SOME(row, maybeRow) {
+      results.push_back(wrapSqlRow(js, kj::mv(row)));
+    } else {
+      break;
+    }
+  }
+
+  return jsg::JsArray(v8::Array::New(js.v8Isolate, results.data(), results.size()));
+}
+
+jsg::JsValue SqlStorage::Cursor::one(jsg::Lock& js) {
+  KJ_IF_SOME(s, state) {
+    cachedColumnNames.ensureInitialized(js, s->query);
+  }
+
+  auto self = JSG_THIS;
+  auto row = JSG_REQUIRE_NONNULL(rowIteratorNext(js, self), Error,
+      "Expected exactly one result from SQL query, but got no results.");
+  auto result = wrapSqlRow(js, kj::mv(row));
+
+  // Manually grab the query, check that there are no more results, and clean it up.
+  auto& query = KJ_ASSERT_NONNULL(state)->query;
+  query.nextRow();
+  bool hadOneResult = query.isDone();
+
+  // Save off row counts before the query goes away, just like iteratorImpl() would do when done.
+  rowsRead = query.getRowsRead();
+  rowsWritten = query.getRowsWritten();
+
+  // End the query even if it wasn't done.
+  state = kj::none;
```

Comment:
I really like that the query gets definitely closed as part of calling `.one()`. Very tidy.